lib/Transforms/HandshakeOptimization.cpp

```diff
@@ -0,0 +1,629 @@
+      mlir::Value conditionBr = condBranchOp.getConditionOperand();
+      mlir::Value dataOperand = condBranchOp.getDataOperand();
+      rewriter.replaceOp(mergeOp, {dataOperand});
+      rewriter.setInsertionPoint(condBranchOp);
+      rewriter.create<handshake::SinkOp>(condBranchOp->getLoc(), conditionBr);
+      rewriter.eraseOp(condBranchOp);
+      return success();
+    }
+    return failure();
+  }
+};
+
+struct RemoveBranchCMergePairs
+    : public OpRewritePattern<handshake::ConditionalBranchOp> {
+  using OpRewritePattern<handshake::ConditionalBranchOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(handshake::ConditionalBranchOp condBranchOp,
+                                PatternRewriter &rewriter) const override {
+    mlir::Operation *useOwner1 = nullptr;
+    mlir::Operation *useOwner2 = nullptr;
+    auto it = condBranchOp.getTrueResult().getUsers().begin();
+    if (it != condBranchOp.getTrueResult().getUsers().end()) {
+      useOwner1 = *it;
+    }
+
+    it = condBranchOp.getFalseResult().getUsers().begin();
+    if (it != condBranchOp.getFalseResult().getUsers().end()) {
+      useOwner2 = *it;
+    }
+
+    if (isa_and_nonnull<handshake::ControlMergeOp>(useOwner1) and
+        useOwner1 == useOwner2) {
+
+      handshake::ControlMergeOp cMergeOp =
+          mlir::cast<handshake::ControlMergeOp>(useOwner1);
+
+      mlir::Value dataOperand = condBranchOp.getDataOperand();
+      mlir::Value conditionBr = condBranchOp.getConditionOperand();
+
+      rewriter.replaceOp(cMergeOp, {dataOperand, conditionBr});
+      rewriter.eraseOp(condBranchOp);
+      return success();
+    }
+    return failure();
+  }
+};
+
+struct RemoveMUXBranchLoopPairs : public OpRewritePattern<handshake::MuxOp> {
+  using OpRewritePattern<handshake::MuxOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(handshake::MuxOp muxOp,
+                                PatternRewriter &rewriter) const override {
+    mlir::Operation *useOwner = nullptr;
+    auto it = muxOp->getUsers().begin();
+    if (it != muxOp->getUsers().end()) {
+      useOwner = *it;
+    }
+    if (isa_and_nonnull<handshake::ConditionalBranchOp>(useOwner)) {
+      handshake::ConditionalBranchOp condBranchOp =
+          mlir::cast<handshake::ConditionalBranchOp>(useOwner);
+      mlir::OperandRange dataOperands = muxOp.getDataOperands();
+      mlir::Value first = dataOperands[0];
+      mlir::Value second = dataOperands[1];
+      mlir::Value resultMux = muxOp.getResult();
+      mlir::Value select = muxOp.getSelectOperand();
+      mlir::Value condition = condBranchOp.getConditionOperand();
+      mlir::Value branchIn = condBranchOp.getDataOperand();
+      mlir::Value trueResult = condBranchOp.getTrueResult();
+      mlir::Value falseResult = condBranchOp.getFalseResult();
+      mlir::ValueRange l = {second, second};
+
+      if (resultMux == branchIn) {
+        if (trueResult == first) {
```

Comment:
This is basically the same piece of code 4 times, and it seems it can be simplified to be made more easily readable.

```cpp
Value srcVal;
if (trueResult == first || trueResult == second)
 srcVal = trueResult;
else if (falseResult == first || falseResult == second)
 srcVal = falseResult
else
 return failure()
Value dstVal = (trueResult == first || falseResult == first) ? second : first;

rewriter.setInsertionPoint(muxOp);
rewriter.create<handshake::SinkOp>(muxOp->getLoc(), select);
rewriter.replaceOp(muxOp, dstVal);
rewriter.replaceAllUsesWith(srcVal, dstVal);
```